### PR TITLE
ENH: Added scipy.signal.tfest for transfer function estimation

### PR DESCRIPTION
**Pull request by fred1608**
#### What does this implement/fix?
This implements a function similar in scope to Matlab's tfest, in which a transfer function can be estimated based on input and output signals (Transfer Function Identification).

#### Additional information
This function is very useful for signal processing and control. It can be used, for example, to identify a process' transfer function in order to design a control system, or model it mathematically.

### CODE REVIEW

**Comment by ilayn** [comment]
Hi @fred1608 

Thanks a lot for your PR. Indeed, tfest is a very useful function. However, as it is now it won't give satisfactory answers if pole zero numbers are increased even slightly. In fact this is one of the hardest problems in identification to get it robust. You can get a feeling why it is from tfest's documentation page on Mathworks site.

**Comment by fred1608** [comment]
Hello @ilayn,
thank you for your reply. I feel this would be indeed a valid addition.
Do you think an R squared metric would suffice, or should there be others as well?

File: THANKS.txt [hunk 1]
```diff
@@ -242,6 +242,7 @@ Mark Borgerding for contributing linalg.convolution_matrix.
 Shashaank N for contributions to scipy.signal.
 Frank Torres for fixing a bug with solve_bvp for large problems.
 Ben West for updating the Gamma distribution documentation.
+Luiz Frederico S S M Villalobos for adding transfer function identification to scipy.signal
```
Comment (rlucas7) [inline THANKS.txt:245]: Can this line be removed from the PR? 
See [thread](https://mail.python.org/pipermail/scipy-dev/2020-August/024331.html) for discussion (changes to this file causes merge conflicts adding effort to contributors and maintainers). 
For now, I've opened a [wiki](https://github.com/scipy/scipy/wiki/*THANKS.txt-additions-modifications-for-1.6.0*)
Once the PR is merged, we add thanks to the wiki. Similar process to what we followed in 1.5.0 release.

**Comment by ilayn** [comment]
I don't think any curve fitting would actually suffice here since the structure of the unknowns is also unknown hence regression won't be possible which is basically the fundamental problem of system identification.

File: THANKS.txt (hunk 1 above)
Comment (fred1608) [reply THANKS.txt:245]: Yes, I'll remove right it away.

**Comment by fred1608** [comment]
@ilayn,
while I agree that the structure is also unknown, I don't see why this would pose a problem for practical implementation. In my experience, transfers functions that have had their order overestimated will create zeros and poles while small coefficients that will have negligible effect on the system. In parallel, functions with underestimated orders will behave as an simplification of their larger counterparts. Naturally, both cases can have an effect on the result, but this is why I spoke about adding a metric to evaluate. That said, this is mostly my anecdotal experience, if you have more information, I would love to hear.

Another thing I would like to add is that matlab's own tfest work in a similar fashion when it comes to defining the number of zeros and poles of the transfer function before identification occurs, so I don't think they directly address this.

**Comment by ilayn** [comment]
You can try yourself. Make a 4 pole/ 3 zero combination system apply a chirp or noise signal. Then add noise on top of the signals and try to curve fit. It will not give satisfactory results via curve fitting. Actually that's why people started to look into system identification in the first place. 

Nonminimal systems actually make things much worse as they try to fit things in the high frequency rollover area which basically you never care.

**Comment by fred1608** [comment]
@ilayn,

I indeed performed the test and it failed as you said. Thank you for your help, I'll try to improve the design and resubmit it at a later stage.

**Comment by rgommers** [comment]
Also a comment on naming: `tfest` is pretty bad (per the usual Matlab "let's choose short, uninformative names"). Can it be named `estimate_tf` or something else more descriptive?

**Comment by fred1608** [comment]
> Also a comment on naming: `tfest` is pretty bad (per the usual Matlab "let's choose short, uninformative names"). Can it be named `estimate_tf` or something else more descriptive?

Sure, I agree, it's probably a better name. I'll change that when I get around to making the other modifications on the code as well.

**Comment by fred1608** [comment]
Hello,
sorry for the long delay in continuing this project.

After some further studies, I've changed the algorithm to one based around least squares. I've also changed the name of the function.

Let me know your opinions on the current system, and if there is anything else required.

**Review by tylerjereddy — COMMENTED** [review]
It looks like something went awry with the rebase here--a bunch of files that shouldn't show changes, like `.mailmap`, are showing changes.

**Comment by fred1608** [comment]
> It looks like something went awry with the rebase here--a bunch of files that shouldn't show changes, like `.mailmap`, are showing changes.

Alright, thanks for the heads up! I'll see what I can do about it.

**Comment by fred1608** [comment]
> It looks like something went awry with the rebase here--a bunch of files that shouldn't show changes, like `.mailmap`, are showing changes.

They should be fixed now. Give me a heads up if you notice anything else that should not be there, or if you have any other inputs for this project.

**Comment by ilayn** [comment]
It seems to me that when you rebased it you didn't force push it so it got all the irrelevant commits together with it. You can see that in the commit history above. There are many ways to revert it ranging from manual to more surgical which you can take whichever you want from the search results https://www.google.com/search?hl=en&q=how%20to%20revert%20a%20rebase

**Comment by rgommers** [comment]
> It seems to me that when you rebased it you didn't force push it so it got all the irrelevant commits together with it. You can see that in the commit history above.

Not force-pushing after a rebase is impossible. I was wondering for a long time how people manage to include 100s of commits in a PR, and recently figured it out after it happened to me. This happens when you do a rebase within a rebase. This can happen if you solve a conflict in the middle of a rebase, and then instead of `git rebase --continue` you do another `git rebase`.

**Comment by ilayn** [comment]
You can actually fall to this scenario after in your branch which is behind the master some commits and ahead at the same time a few commits. Then you do a `git rebase master` then you are in a situation, (in posh-git terminology) even when there is no conflict , you'll get

```
[mybranch ↓3 ↑123]
```

meaning it will pull the existing three commits and submit the 120 commits that it was behind then will place the 3 commits at the top. That's what you see here. 

If you force push it, it will overwrite the remote and the commits will be synced with the more recent master showing only the contributed commits.

**Comment by ev-br** [comment]
> I was wondering for a long time how people manage to include 100s of commits in a PR

One other way --- fairly common in my experience --- to include irrelevant commits is to do `$ git rebase -i <very_old_commit>`. Apparently many people expect do not expect that git changes hashes of all commits in the rebase, even if they are not changed.
(which is a plausible assumption, but simply not how git works)

**Comment by fred1608** [comment]
Thanks guys, I wasn't aware of this behavior. I'll do some research and see that the commit history is fixed.

Just to be sure, the problem is only with the history, rights? Or have I merged with some files that I shouldn't?

**Comment by roryyorke** [comment]
There was  a prior attempt at this: https://github.com/scipy/scipy/pull/5737.  We've also a request over at [python-control](https://github.com/python-control/python-control/issues/550) to implement Matlab's `fitfrd`, which is similar, but frequency-domain only.  It turns out SLICOT (Fortran, recently BSD-licensed) does implement this in `SB10YD`, but we haven't made it available in Slycot.

**Comment by ilayn** [comment]
Since there is not much interest in subsuming slicot here or overhauling signal.lti , I think pointing to python-control is the sane option for this feature.

**Comment by lucascolley** [comment]
Hi @fred1608, the rebase problem is still here, and we have a -1 on inclusion in SciPy, so I'm going to close this.

If you would like to return to this and remake the PR, feel free to start a discussion on the mailing list, although it looks like there isn't much interest.

**Pull request closed without merging**